### DISCUSSION TITLE
MOTOR CONTROL NOT RUNNING

### DISCUSSION THREAD

**Discussion by thanhtruongphan** [fjp/2wd-robot#95] [Q&A]
I run with command line ```roslaunch diffbot_bringup bringup.launch```
then I turn 2 wheels and the encoder-ticks woking proberly.
then I ```rosrun rqt_robot_steering rqt_robot_steering``` and the _**wheel_cmd_velocities**_ read ok, too.

but the two wheels not move.
@fjp hiii... did i do something wrong or is there a mistake in the repositories code...? can you have any ideas, 
thank you so much and have a nice day~ 👍 :D

![image](https://github.com/user-attachments/assets/012048d0-57ad-40ec-9157-0afa541329fd)

**Comment by thanhtruongphan** [comment]
[here](https://github.com/[thanhtruongphan/diff-bot-mod-with-l298](https://github.com/thanhtruongphan/diff-bot-mod-with-l298)) is the code, I only edited the **_base_controller_** part to suit the use with L298, the rest is kept the same as the original code.

**Reply by fjp** [reply]
Hi @thanhtruongphan thanks for the update using a teensy. Can you please post another link or edit the first one? I only get 404 with the current repo link

**Reply by thanhtruongphan** [reply]
yeah, [here](https://github.com/thanhtruongphan/diff-bot-mod-with-l298) is the new update link.

I just modified only [base_controller](https://github.com/thanhtruongphan/diff-bot-mod-with-l298/tree/master/diffbot_base/scripts/base_controller)

> thanks for the update using a teensy
hah i feel low mood while it's still not running and have a lot of error, I'll come back to stm32 when I feel better... 😂

**Comment by fjp** [comment]
Thanks for the new code link. I've reviewed it and everything looks like it should work, so nothing obviously wrong with it in my opinion. What's strange is that your log output shows `motor_cmd_left/right_=0` even though `pid_left/righ_errors are **not** equal to zero. 

It seems that `motor_pid_left/right.compute()` returns zero in the `write` method:

```cpp
template <typename L298MotorController>
void diffbot::BaseController<L298MotorController>::write()
{
    motor_cmd_left_ = motor_pid_left_.compute(wheel_cmd_velocity_left_, joint_state_left_.angular_velocity_);
    motor_cmd_right_ = motor_pid_right_.compute(wheel_cmd_velocity_right_, joint_state_right_.angular_velocity_);
    
    p_motor_controller_left_->setSpeed(motor_cmd_left_);
    p_motor_controller_right_->setSpeed(motor_cmd_right_);
    
}
```

Looking at your debug output, I can see:

1. `wheel_cmd_velocity_left_` = 7.67 (this would be your setpoint coming from the high level hardware interface from the Raspi SBC, which is good)
2. `measured_ang_vel_left` = 0.00 (this is your measured_value)
3. `pid_left_errors` (p, i, d) shows 7.67 which means:
    - `proportional_` = 7.67 (this is correct since error = setpoint - measured = 7.67 - 0.00)
4. `motor_cmd_left_ = 0` (this is your final pid output after `constrain()`)


With kp_ = 10 and proportional_ = 7.67, the PID calculation should be:

```
pid = 10 * 7.67 = 76.7
```


Your `min_val_, max_val_` ([`PWM_MIN, PWM_MAX`](https://github.com/thanhtruongphan/diff-bot-mod-with-l298/blob/035d2e5e69cc2c61c63a6309c6ba4eb11cf56182/diffbot_base/scripts/base_controller/lib/config/diffbot_base_config.h#L39-L40)) as well as your [PID values](https://github.com/thanhtruongphan/diff-bot-mod-with-l298/blob/035d2e5e69cc2c61c63a6309c6ba4eb11cf56182/diffbot_base/scripts/base_controller/lib/config/diffbot_base_config.h#L24-L26) seem to be good as well. Although I think `K_P` of 10 will be too high. I guess you have tried already the defaults?

```cpp
#define K_P 0.6 // P constant
#define K_I 0.3 // I constant
#define K_D 0.5 // D constant
```


I'd suggest to add more logging in the `PID` class by either
1. [preferred logging approach] pass the ROS node handle to the `PID` class [here](https://github.com/thanhtruongphan/diff-bot-mod-with-l298/blob/035d2e5e69cc2c61c63a6309c6ba4eb11cf56182/diffbot_base/scripts/base_controller/lib/base_controller/base_controller.h#L155C1-L156C56) after adding it as a constructor argument and having it as a member (similar to how its done in [Encoder class](https://github.com/thanhtruongphan/diff-bot-mod-with-l298/blob/035d2e5e69cc2c61c63a6309c6ba4eb11cf56182/diffbot_base/scripts/base_controller/lib/encoder/encoder_diffbot.h#L45)). Then use it to log [`pid`](https://github.com/thanhtruongphan/diff-bot-mod-with-l298/blob/035d2e5e69cc2c61c63a6309c6ba4eb11cf56182/diffbot_base/scripts/base_controller/lib/pid/pid.cpp#L29) just before the Arduino `constrain()` method and also log the `k_p` value (which should be 10.0). You could also add another local variable, e.g. `constrain_value = constrain(pid, min_val_, max_val_)` and log it as well before returning it.
2. [less preferred logging approach] or turn the [local `pid` variable](https://github.com/thanhtruongphan/diff-bot-mod-with-l298/blob/035d2e5e69cc2c61c63a6309c6ba4eb11cf56182/diffbot_base/scripts/base_controller/lib/pid/pid.cpp#L16) into a member variable in [pid.h](https://github.com/thanhtruongphan/diff-bot-mod-with-l298/blob/035d2e5e69cc2c61c63a6309c6ba4eb11cf56182/diffbot_base/scripts/base_controller/lib/pid/pid.h#L18) and then log its value in the [`printDebug()` method](https://github.com/thanhtruongphan/diff-bot-mod-with-l298/blob/035d2e5e69cc2c61c63a6309c6ba4eb11cf56182/diffbot_base/scripts/base_controller/lib/base_controller/base_controller.h#L323-L327). I'd suggest to also log `k_p_` and you could also do the same for additional intermediate variables, like `constrain_value_`.

Looking forward to your new log output 😊

**Reply by thanhtruongphan** [reply]
thank you, @fjp :1st_place_medal: 

i probably won't be able to use teensy anymore, my board just broke this morning 🥹 I was still using it yesterday but this morning, when i plugged in the USB power source, the laptop didn't recognize the USB port, i suddenly realized that the power light wasn't on anymore. maybe it's broken. i don't know, why did such an expensive board break, while Arduino and STM32 were fine. hah first time i had a circuit break like that. maybe I'll study the integration with stm32 again soon... and, I also try to so something with the andino that you have mentioned before.

:1st_place_medal: 😊

**Reply by fjp** [reply]
That's sad to hear that your Teensy broke 😢 If it's within the warranty period, it's definitely worth reaching out to where you purchased it for a potential replacement. Or you can contact PJRC through their forum (https://forum.pjrc.com/) or email Paul Stoffregen (the creator) directly - they're known for good customer service. Even if it's outside the warranty period, you might want to contact them anyway - sometimes they can offer advice or discounted replacements, especially if it's an unexpected failure.

You might want to try a few troubleshooting steps:

- Try a different USB cable - sometimes it's just a cable issue
- Check if there are any visible damage to the USB connector or board
- Try a different USB port on your laptop
- If possible, test with a different computer

> maybe I'll study the integration with stm32 again soon

If it's truly dead, it would be great to see ROS serial working on STM32! But I agree it has a bit steeper learning curve than Teensy/Arduino.

> and, I also try to so something with the andino that you have mentioned before.

Yes, the Andino software should also work with your Ardunio board. In case you plan to stick with ROS 1 and want to use diffbot software you'd have to edit the diffbot_base high level hardware interface running on the Raspi to use [libserial](https://libserial.readthedocs.io/en/latest/index.html), as [explained previously](https://github.com/orgs/ros-mobile-robots/discussions/94#discussioncomment-10836028).

**Reply by thanhtruongphan** [reply]
thanks for your advice, @fjp 

I'm come from Vietnam and no PJRC official on my region. The shop where I bought, it doesn't waranty for electronic components, electronic circuits like Teensy :smile: 

It would be a hassle if I contacted Teensy service and got it under warranty, if I lucky...but never mind, I'll deal with it later.

**Comment by thanhtruongphan** [comment]
but, I wonder, will Teensy 3.2 work well with this diffbot?

If yes, why does stm32f411 often get "lost sync..."
I tested, if comment out all "debug info" and "loginfo" then there is no "lost sync..." error. 

I thought it was because the stm32 hardware is much smaller than the Teensy 4.0, but in "list-parts" there are two options "Teensy 4.0 or Teensy 3.2"
>>>    ![image](https://github.com/user-attachments/assets/a0af3e84-a8ae-43fe-8cf0-66b2e6236a5e)

STM32F411CEU6 (128kb SRAM, 512kb FLASH, 100MHz)
Teensy 3.2             (64kb SRAM, 256kb FLASH, 72MHz)

"wrong checksum for topic id..." appears on Teensy 4.0 but the system still works fine. So temporarily don't worry about it.

when launching "diffbot_bringup", the first few seconds will "Emmergency STOP" until the connection to the ROS system is recognized. It will print out the "debug" parameters and the system will start working.

Next is the "```log output shows motor_cmd_left/right_=0 even though `pid_left/righ_errors are not equal to zero```" issue as we mentioned above. I will need to deal with it when the above part (stm32 instead Teensy) is ok.
